nca/FWRules/ConnectivityGraph.py

```diff
@@ -133,26 +151,139 @@ def _creates_cliqued_graph(directed_edges):
+        """
+        A biclique is a pair of sets. each node at the firs set as an edge to each node in the second set.
+        The algorithm:
+        1. for each src_node, find the set of the destination nodes (i.e. dst_sets).
+        2. for each dst_set found in (1), find the set of sources that connected to every node in the dst_set (i.e. src_set).
+           the (src_set, dst_set) is a biclique
+        3. find the best biclique - the biclique that reduce the highest number of edges
+        4. remove the biclique from the graph, and go back to (1) till no biclique found
 
-        return directed_edges, not_directed_edges, cliques_nodes
+        :param: directed_edges: list of pairs representing the original graph
+        return: truple( list, list, list)
+        list: list of the directed edges that was left in the graph
+        list: list the new bicliques nodes
+        list: list the new bicliques edged
+
+        """
+        bicliques_nodes = []
+        bicliques_edges = set()
+
+        while directed_edges:
+            # find dst_sets:
+            all_sources = set(edge[0] for edge in directed_edges)
+            src_to_dst_set = {src: frozenset(e[1] for e in directed_edges if e[0] == src) for src in all_sources}
+            all_dst_set = frozenset(src_to_dst_set.values())
+
+            # find all bicliques:
+            all_bicliques = []
+            for dst_set in all_dst_set:
+                src_set = frozenset(src for src in all_sources if src_to_dst_set[src] >= dst_set)
+                all_bicliques.append((src_set, dst_set))
+            # find the best biclique:
+            bicliques_ranks = {(s, d): len(s) * len(d) - len(s) - len(d) for s, d in all_bicliques}
+            if max(bicliques_ranks.values()) < 1:
+                break
+            best_biclique = max(bicliques_ranks, key=bicliques_ranks.get)
+
+            # create the new biclique, and remove it from the graph:
+            directed_edges -= set(itertools.product(best_biclique[0], best_biclique[1]))
+            biclique_name = f'biclique_{conn_str}{len(bicliques_nodes)}'
+            biclique_node = (biclique_name, '')
+            bicliques_nodes.append(biclique_node)
+            bicliques_edges |= set((src, biclique_node) for src in best_biclique[0])
+            bicliques_edges |= set((biclique_node, dst) for dst in best_biclique[1])
+        return directed_edges, bicliques_edges, bicliques_nodes
+
+    @staticmethod
+    def _find_equal_groups(peers_edges):
+        """
+        find set of peers that have the same io edges and the same namespace
+        :param: peers_edges: dict peer -> [peer edges]
+        return: list
+        list: list of lists, each list is a group of peers
+        list: list of peers that are not in any group
+        """
+        groups_dict = defaultdict(list)
+        for peer, peer_edges in peers_edges.items():
+            groups_dict[(peer.namespace, peer_edges)].append(peer)
```

Comment:
We probably don't want to add to `groups_dict` peers for which `peer_edges` is empty. Currently this seems to break testing.

Reply:
we do want to add empty peer edges, all peers with no edges are grouped together to one group.
in this case, the group will not have a self loop.
so I fixed it in line 278